File: android/app/src/main/java/com/vegettable/app/ui/map/MapActivity.java
package com.vegettable.app.ui.map;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.vegettable.app.R;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class MapActivity extends AppCompatActivity {

    private MarketAdapter adapter;
    private String selectedRegion = null;

    // 批發市場資料
    private static final List<MarketItem> ALL_MARKETS = Arrays.asList(
        new MarketItem("台北一", "台北市萬華區萬大路533號", "北部", 25.0258, 121.5010),
        new MarketItem("台北二", "台北市中山區民族東路336號", "北部", 25.0690, 121.5375),
        new MarketItem("三重", "新北市三重區大同北路107號", "北部", 25.0620, 121.4872),
        new MarketItem("桃園", "桃園市桃園區中山路590號", "北部", 24.9917, 121.3125),
        new MarketItem("台中", "台中市西屯區中清路350號", "中部", 24.1795, 120.6547),
        new MarketItem("溪湖", "彰化縣溪湖鎮彰水路四段510號", "中部", 23.9617, 120.4793),
        new MarketItem("西螺", "雲林縣西螺鎮九隆里延平路248號", "中部", 23.7983, 120.4602),
        new MarketItem("嘉義", "嘉義市西區博愛路二段459號", "南部", 23.4817, 120.4343),
        new MarketItem("台南", "台南市北區忠北街7號", "南部", 23.0125, 120.2153),
        new MarketItem("鳳山", "高雄市鳳山區建國路三段39號", "南部", 22.6273, 120.3419),
        new MarketItem("屏東", "屏東縣屏東市工業路9號", "南部", 22.6656, 120.4950),
        new MarketItem("宜蘭", "宜蘭縣宜蘭市環市東路二段1號", "東部", 24.7469, 121.7515),
        new MarketItem("花蓮", "花蓮縣花蓮市中華路100號", "東部", 23.9872, 121.6044)
    );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_map);

        findViewById(R.id.btn_back).setOnClickListener(v -> finish());

        RecyclerView rvMarkets = findViewById(R.id.rv_markets);
        rvMarkets.setLayoutManager(new LinearLayoutManager(this));
        adapter = new MarketAdapter();
        rvMarkets.setAdapter(adapter);
        adapter.setItems(ALL_MARKETS);

        setupRegionChips();
    }

    private void setupRegionChips() {
        ChipGroup group = findViewById(R.id.chip_group_region);
        String[] regions = {"全部", "北部", "中部", "南部", "東部"};
        for (String r : regions) {
            Chip chip = new Chip(this);
            chip.setText(r);
            chip.setCheckable(true);
            if ("全部".equals(r)) chip.setChecked(true);
            chip.setOnCheckedChangeListener((btn, checked) -> {
                if (checked) {
                    selectedRegion = "全部".equals(btn.getText().toString()) ? null : btn.getText().toString();
                    filterMarkets();
                }
            });
            group.addView(chip);
        }
    }

    private void filterMarkets() {
        if (selectedRegion == null) {
            adapter.setItems(ALL_MARKETS);
        } else {
            List<MarketItem> filtered = new ArrayList<>();
            for (MarketItem m : ALL_MARKETS) {
                if (selectedRegion.equals(m.region)) {
                    filtered.add(m);
                }
            }
            adapter.setItems(filtered);
        }
    }

    // ─── Data ────────────────────────────────────────────────
    static class MarketItem {
        final String name, address, region;
        final double lat, lng;
        MarketItem(String name, String address, String region, double lat, double lng) {
            this.name = name; this.address = address; this.region = region;
            this.lat = lat; this.lng = lng;
        }
    }

    // ─── Adapter ────────────────────────────────────────────
    class MarketAdapter extends RecyclerView.Adapter<MarketAdapter.VH> {
        private List<MarketItem> items = new ArrayList<>();

        void setItems(List<MarketItem> items) {
            this.items = items;
            notifyDataSetChanged();
        }

        @NonNull @Override
        public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_market, parent, false);
            return new VH(v);
        }

        @Override
        public void onBindViewHolder(@NonNull VH h, int pos) {
            MarketItem m = items.get(pos);
            h.tvName.setText(m.name + "果菜批發市場");
            h.tvAddress.setText(m.address);
            h.tvRegion.setText(m.region);
            h.btnNavigate.setOnClickListener(v -> {
                Uri uri = Uri.parse("geo:" + m.lat + "," + m.lng + "?q=" + Uri.encode(m.address));
                Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                startActivity(intent);
            });
        }

        @Override public int getItemCount() { return items.size(); }

        class VH extends RecyclerView.ViewHolder {
            TextView tvName, tvAddress, tvRegion;
            MaterialButton btnNavigate;
            VH(View v) {
                super(v);
                tvName = v.findViewById(R.id.tv_market_name);
                tvAddress = v.findViewById(R.id.tv_address);
                tvRegion = v.findViewById(R.id.tv_region);
                btnNavigate = v.findViewById(R.id.btn_navigate);
            }
        }
    }
}

File: android/app/src/main/java/com/vegettable/app/ui/map/MapActivity.kt
package com.vegettable.app.ui.map

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.button.MaterialButton
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import com.vegettable.app.R
import com.vegettable.app.model.ApiResponse
import com.vegettable.app.model.Market
import com.vegettable.app.network.ApiClient
import org.osmdroid.config.Configuration
import org.osmdroid.tileprovider.tilesource.TileSourceFactory
import org.osmdroid.util.GeoPoint
import org.osmdroid.views.MapView
import org.osmdroid.views.overlay.Marker
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MapActivity : AppCompatActivity() {
    private var adapter: MarketAdapter? = null
    private var allMarkets: MutableList<Market> = mutableListOf()
    private var selectedRegion: String? = null
    private var showingMap = false

    private lateinit var mapView: MapView
    private lateinit var rvMarkets: RecyclerView
    private lateinit var btnToggle: MaterialButton

    override fun onCreate(savedInstanceState: Bundle?) {
        // osmdroid 快取設定（必須在 super 之前）
        Configuration.getInstance().load(applicationContext,
            androidx.preference.PreferenceManager.getDefaultSharedPreferences(applicationContext))
        Configuration.getInstance().userAgentValue = packageName

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_map)

        findViewById<View>(R.id.btn_back).setOnClickListener { finish() }

        mapView = findViewById(R.id.map_view)
        rvMarkets = findViewById(R.id.rv_markets)
        btnToggle = findViewById(R.id.btn_toggle_view)

        setupMap()
        setupRecyclerView()
        setupRegionChips()
        setupToggleButton()
        loadMarketsFromApi()
    }

    private fun setupMap() {
        mapView.setTileSource(TileSourceFactory.MAPNIK)
        mapView.setMultiTouchControls(true)
        mapView.controller.setZoom(8.0)
        // 台灣中心點
        mapView.controller.setCenter(GeoPoint(23.6, 120.9))
    }

    private fun setupRecyclerView() {
        rvMarkets.layoutManager = LinearLayoutManager(this)
        adapter = MarketAdapter()
        rvMarkets.adapter = adapter
    }

    private fun setupToggleButton() {
        btnToggle.setOnClickListener {
            showingMap = !showingMap
            if (showingMap) {
                mapView.visibility = View.VISIBLE
                rvMarkets.visibility = View.GONE
                btnToggle.text = "列表"
                addMapMarkers(filteredMarkets())
            } else {
                mapView.visibility = View.GONE
                rvMarkets.visibility = View.VISIBLE
                btnToggle.text = "地圖"
            }
        }
    }

    private fun loadMarketsFromApi() {
        ApiClient.instance?.api?.markets
            ?.enqueue(object : Callback<ApiResponse<MutableList<Market?>?>?> {
                override fun onResponse(
                    call: Call<ApiResponse<MutableList<Market?>?>?>,
                    response: Response<ApiResponse<MutableList<Market?>?>?>
                ) {
                    if (response.isSuccessful && response.body()?.isSuccess == true) {
                        val markets = response.body()?.data?.filterNotNull()?.toMutableList()
                            ?: mutableListOf()
                        allMarkets = markets
                        filterMarkets()
                    } else {
                        Toast.makeText(this@MapActivity, "市場資料載入失敗", Toast.LENGTH_SHORT).show()
                    }
                }

                override fun onFailure(call: Call<ApiResponse<MutableList<Market?>?>?>, t: Throwable) {
                    Toast.makeText(this@MapActivity, "網路連線異常", Toast.LENGTH_SHORT).show()
                }
            })
    }

    private fun setupRegionChips() {
        val group = findViewById<ChipGroup>(R.id.chip_group_region)
        val regions = arrayOf("全部", "北部", "中部", "南部", "東部")
        for (r in regions) {
            val chip = Chip(this)
            chip.text = r
            chip.isCheckable = true
            if ("全部" == r) chip.isChecked = true
            chip.setOnCheckedChangeListener { btn: CompoundButton, checked: Boolean ->
                if (checked) {
                    selectedRegion = if ("全部" == btn.text.toString()) null else btn.text.toString()
                    filterMarkets()
                }
            }
            group.addView(chip)
        }
    }

    private fun filteredMarkets(): MutableList<Market> {
        return if (selectedRegion == null) {
            allMarkets.toMutableList()
        } else {
            allMarkets.filter { it.region == selectedRegion }.toMutableList()
        }
    }

    private fun filterMarkets() {
        val filtered = filteredMarkets()
        adapter?.setItems(filtered)
        if (showingMap) {
            addMapMarkers(filtered)
        }
    }

    private fun addMapMarkers(markets: List<Market>) {
        mapView.overlays.clear()
        for (m in markets) {
            val lat = m.latitude ?: continue
            val lng = m.longitude ?: continue
            if (lat == 0.0 && lng == 0.0) continue

            val marker = Marker(mapView)
            marker.position = GeoPoint(lat, lng)
            marker.title = (m.marketName ?: "") + "果菜批發市場"
            marker.snippet = m.address ?: ""
            marker.setAnchor(Marker.ANCHOR_CENTER, Marker.ANCHOR_BOTTOM)
            marker.setOnMarkerClickListener { mk, _ ->
                mk.showInfoWindow()
                true
            }
            mapView.overlays.add(marker)
        }

        // 有市場時縮放到合適範圍
        if (markets.isNotEmpty()) {
            val validMarkets = markets.filter {
                (it.latitude ?: 0.0) != 0.0 || (it.longitude ?: 0.0) != 0.0
            }
            if (validMarkets.size == 1) {
                val m = validMarkets.first()
                mapView.controller.animateTo(GeoPoint(m.latitude!!, m.longitude!!))
                mapView.controller.setZoom(14.0)
            } else if (validMarkets.size > 1) {
                mapView.controller.setZoom(8.0)
                mapView.controller.setCenter(GeoPoint(23.6, 120.9))
            }
        }
        mapView.invalidate()
    }

    override fun onResume() {
        super.onResume()
        mapView.onResume()
    }

    override fun onPause() {
        super.onPause()
        mapView.onPause()
    }

    override fun onDestroy() {
        super.onDestroy()
        mapView.onDetach()
    }

    // ─── Adapter ────────────────────────────────────────────
    internal inner class MarketAdapter : RecyclerView.Adapter<MarketAdapter.VH>() {
        private var items: MutableList<Market> = mutableListOf()

        fun setItems(newItems: MutableList<Market>) {
            val diff = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize() = items.size
                override fun getNewListSize() = newItems.size
                override fun areItemsTheSame(o: Int, n: Int) = items[o].marketCode == newItems[n].marketCode
                override fun areContentsTheSame(o: Int, n: Int) = items[o].marketName == newItems[n].marketName
            })
            items = newItems
            diff.dispatchUpdatesTo(this)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
            val v = LayoutInflater.from(parent.context).inflate(R.layout.item_market, parent, false)
            return VH(v)
        }

        override fun onBindViewHolder(h: VH, pos: Int) {
            val m = items[pos]
            h.tvName.text = (m.marketName ?: "") + "果菜批發市場"
            h.tvAddress.text = m.address ?: ""
            h.tvRegion.text = m.region ?: ""
            h.btnNavigate.setOnClickListener {
                val address = m.address ?: return@setOnClickListener
                val uri = Uri.parse("geo:${m.latitude},${m.longitude}?q=${Uri.encode(address)}")
                startActivity(Intent(Intent.ACTION_VIEW, uri))
            }
        }

        override fun getItemCount() = items.size

        inner class VH(v: View) : RecyclerView.ViewHolder(v) {
            val tvName: TextView = v.findViewById(R.id.tv_market_name)
            val tvAddress: TextView = v.findViewById(R.id.tv_address)
            val tvRegion: TextView = v.findViewById(R.id.tv_region)
            val btnNavigate: MaterialButton = v.findViewById(R.id.btn_navigate)
        }
    }
}
